### ISSUE TITLE
Filtering rules, result types doesn't work as expected

### ISSUE THREAD

**Issue by mohanraj-r** [dequelabs/axe-core#1540]
Expectation: 
* Disabling a rule should prevent that rule from running. 
* Filtering by result types should present only the selected result type in output. 

Actual: 
* The rule seems to be still run (as per performanceTimer logs). `axe.run()` take the same amount of time with the rule disabled, even though the rule takes non-trivial amount of time.
* result types other than the one specified are also present in output.

Motivation: Ability to turn-off selected rules  for large pages #1464 and result types to improve performance as documented in https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#use-resulttypes

<pre><code>
axe-core version: 3.2.2

Browser and Assistive Technology versions
Google chrome: Version 73.0.3683.103
</code></pre>

```
await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: false},
   },
  "resultTypes": ["violations"],
  "performanceTimer": true
})
```

Output:
```
PerformanceTimer reports the rule being run. The total time taken for execution is the same as with the rule enabled.
...
Measure rule_color-contrast took 5298.990000009537ms
15:11:51.155 axe.min.js:12 
Measure runchecks_color-contrast took 4943.65499997139ms
...
Measure axe took 14006ms
```

Other result types such as `inapplicable`, `passes` etc are still reported even though `"resultTypes": ["violations"]` is specified
```
{inapplicable: (28) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]incomplete: [{…}]passes: (34) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]testEngine: {name: "axe-core", version: "3.2.2"}testEnvironment: {userAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_3) Ap…L, like Gecko) Chrome/73.0.3683.103 Safari/537.36", windowWidth: 960, windowHeight: 120, orientationAngle: 90, orientationType: "portrait-secondary"}testRunner: {name: "axe"}timestamp: "2019-05-02T22:11:52.898Z"toolOptions: {rules: {…}, resultTypes: Array(1), performanceTimer: true, reporter: "v1"} url: "file:///tmp/test.html"violations: (11) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]constructor: (...)hasOwnProperty: (...)isPrototypeOf: (...)propertyIsEnumerable: (...)toLocaleString: (...)toString: (...)valueOf: (...)__defineGetter__: (...)__defineSetter__: (...)__lookupGetter__: (...)__lookupSetter__: (...)__proto__: Object
```

**Comment by mohanraj-r** [comment]
Also tested with the latest changes in develop (b573b1c1e554e3e93e53716e95e92ed3f4acd4f1) and see the same issues.

**Comment by straker** [comment]
I'm not seeing the `color-contrast` rule being run when `{ color-contrast: { enabled: false } }` is set (perf metric logs in the console don't show a measure for it) using axe v3.2.2 and I believe `resultTypes` is working as designed.

According to the docs

> An approach you can take to reducing the time is use the resultTypes option. By calling axe.run with the following options, axe-core will only return the full details of the violations array and will only return one instance of each of the inapplicable, incomplete and pass arrays for each rule that has at least one of those entries. This will reduce the amount of computation that axe-core does for the unique selectors.

The `resultTypes` option only forces the rule to list at most 1 node for the rule and not all nodes. You can see this if you run on https://www.tensorflow.org/api_docs/python/tf/InteractiveSession#__init__. Without the `resultTypes` option there are 63 nodes that fail the `button-name` rule, but if we only look at `passes` there is only 1 node listed. This helps speed up the end part of axe that generates unique selectors for each node listed in the results.

```js
var results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: false},
   }
})
console.log(results.violations[1].nodes.length)  // => 63

results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: false},
   },
  "resultTypes": ["passes"]
})
console.log(results.violations[1].nodes.length)  // => 1
```

**Comment by mohanraj-r** [comment]
@straker @WilcoFiers thanks for validating 

I tested this again to make sure. What I am finding is
1. On a fresh load, disabling a rule works as expected, i.e. with `"color-contrast": {enabled: false},` I am not seeing any performance timer logs related to `color-contrast`
```
var results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: false},
   },
  "resultTypes": ["violations"],
  "performanceTimer": true
})
...
# no color-contrast in timer logs
```

2. But after running axe with the rule enabled once, disabling it subsequently doesn't seem to take effect. 
```
var results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: true},
   },
  "resultTypes": ["violations"],
  "performanceTimer": true
})
...
12:57:17.884 Measure rule_color-contrast took 562.2900000307709ms axe.js:3279 
12:57:17.885 Measure runchecks_color-contrast took 269.4600000977516ms axe.js:3279 
```

Running it again with enabled false in the same session results in color-contrast being shown in perf timer logs
```
var results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: false},
   },
  "resultTypes": ["violations"],
  "performanceTimer": true
})
...
13:01:15.895 Measure rule_color-contrast took 562.2900000307709ms
13:01:15.896 Measure runchecks_color-contrast took 269.4600000977516ms
...
```

**Comment by dylanb** [comment]
@straker this could in fact be a bug when running multiple times with options.

**Comment by straker** [comment]
I think it's just leftover `performance.measures()` that axe is printing out. If you clear the marks and measures before you run again, the logs don't show it being run.

```js
var results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: true},
   },
  "resultTypes": ["violations"],
  "performanceTimer": true
})

performance.clearMarks(); 
performance.clearMeasures();
var results = await axe.run(document, options={
  "rules": {
     "color-contrast": {enabled: false},
   },
  "resultTypes": ["violations"],
  "performanceTimer": true
})
```
